test/libsolutil/CommonIO.cpp

```diff
@@ -98,6 +118,426 @@ BOOST_AUTO_TEST_CASE(readBytes_past_end)
+	BOOST_TEST(absolutePath("...a", reference) == "...a");
+
+	BOOST_TEST(absolutePath("/./", reference) == "/./");
+	BOOST_TEST(absolutePath("/././", reference) == "/././");
+	BOOST_TEST(absolutePath("/../", reference) == "/../");
+	BOOST_TEST(absolutePath("/../../", reference) == "/../../");
+	BOOST_TEST(absolutePath("//..//..//", reference) == "//..//..//");
+	BOOST_TEST(absolutePath("///..///..///", reference) == "///..///..///");
+
+	BOOST_TEST(absolutePath("@", reference) == "@");
+	BOOST_TEST(absolutePath(":", reference) == ":");
+	BOOST_TEST(absolutePath("|", reference) == "|");
+	BOOST_TEST(absolutePath("<>", reference) == "<>");
+	BOOST_TEST(absolutePath("123", reference) == "123");
+	BOOST_TEST(absolutePath("https://example.com", reference) == "https://example.com");
+
+	BOOST_TEST(absolutePath("a", reference) == "a");
+	BOOST_TEST(absolutePath("a/", reference) == "a/");
+	BOOST_TEST(absolutePath("/a", reference) == "/a");
+	BOOST_TEST(absolutePath("/a/", reference) == "/a/");
+
+	BOOST_TEST(absolutePath("a/b", reference) == "a/b");
+	BOOST_TEST(absolutePath("a/b/", reference) == "a/b/");
+	BOOST_TEST(absolutePath("/a/b", reference) == "/a/b");
+	BOOST_TEST(absolutePath("/a/b/", reference) == "/a/b/");
+
+	BOOST_TEST(absolutePath("/.", reference) == "/.");
+	BOOST_TEST(absolutePath("a.", reference) == "a.");
+	BOOST_TEST(absolutePath("a/.", reference) == "a/.");
+	BOOST_TEST(absolutePath("/a/.", reference) == "/a/.");
+	BOOST_TEST(absolutePath("a/./", reference) == "a/./");
+	BOOST_TEST(absolutePath("/a/./", reference) == "/a/./");
+
+	BOOST_TEST(absolutePath("/..", reference) == "/..");
+	BOOST_TEST(absolutePath("a..", reference) == "a..");
+	BOOST_TEST(absolutePath("a/..", reference) == "a/..");
+	BOOST_TEST(absolutePath("/a/..", reference) == "/a/..");
+	BOOST_TEST(absolutePath("a/../", reference) == "a/../");
+	BOOST_TEST(absolutePath("/a/../", reference) == "/a/../");
+
+	BOOST_TEST(absolutePath("a//", reference) == "a//");
+	BOOST_TEST(absolutePath("//a", reference) == "//a");
+	BOOST_TEST(absolutePath("//a//", reference) == "//a//");
+
+	BOOST_TEST(absolutePath("a//b", reference) == "a//b");
+	BOOST_TEST(absolutePath("a//b//", reference) == "a//b//");
+	BOOST_TEST(absolutePath("//a//b", reference) == "//a//b");
+	BOOST_TEST(absolutePath("//a//b//", reference) == "//a//b//");
+
+	BOOST_TEST(absolutePath("a\\b", reference) == "a\\b");
+	BOOST_TEST(absolutePath("a\\b\\", reference) == "a\\b\\");
+	BOOST_TEST(absolutePath("\\a\\b", reference) == "\\a\\b");
+	BOOST_TEST(absolutePath("\\a\\b\\", reference) == "\\a\\b\\");
+}
+
+BOOST_DATA_TEST_CASE(
+	absolutePath_relative_import_no_backtracking_into_reference,
+	boost::unit_test::data::make(std::vector<std::tuple<char const*, std::string>>{
+		{"", ""},
+		{"/", "/"},
+		{"x", ""},
+		{"/x", "/"},
+		{"x/", "x"},
+		{"/x/", "/x"},
+		{"x/y", "x"},
+		{"/x/y", "/x"},
+		{"x/y/", "x/y"},
+		{"/x/y/", "/x/y"},
+	}),
+	reference,
+	parent
```

Comment:
I gave up on looking this up in the boost docs, so I'll just ask here - what are `reference` and `parent`? First and second element of the tuple?

Reply:
This has nothing to do with boost :) The fact that `absolutePath()` uses boost in the first place is highly misleading (and wrong). Those are source unit names, not paths.

The docs you need here is actually this: [Relative Imports](https://docs.soliditylang.org/en/latest/path-resolution.html#relative-imports).

Here "reference" is the source unit name of the importing file. "parent" is just the parent dir of the expected result. I factored it out of the test because in this test scenario it should never change - it's only the final part of the resulting source unit name that changes.